turbopack/crates/turbo-tasks-macros/src/func.rs

```diff
@@ -109,61 +108,56 @@ impl TurboFn<'_> {
+                                        "{} cannot take self by mutable reference, use &self or \
+                                         self: Vc<Self> instead",
+                                        definition_context.function_type(),
+                                    ))
+                                    .emit();
+                                return None;
+                            }
 
-                    match &reference {
-                        None => {
-                            receiver
-                                .span()
-                                .unwrap()
-                                .error(format!(
-                                    "{} cannot take self by value, use &self or self: Vc<Self> \
-                                     instead",
-                                    definition_context.function_type(),
-                                ))
-                                .emit();
-                            return None;
+                            match type_reference.elem.as_ref() {
+                                syn::Type::Path(TypePath { qself: None, path })
+                                    if path.is_ident("Self") => {}
+                                _ => {
+                                    self_type
+                                        .span()
+                                        .unwrap()
+                                        .error(
+                                            "Unexpected `self` type, use `&self` or `self: \
+                                             Vc<Self>",
+                                        )
+                                        .emit();
+                                    return None;
+                                }
+                            }
                         }
-                        Some((_, Some(lifetime))) => {
-                            lifetime
+                        syn::Type::Path(_) => {}
+                        _ => {
+                            self_type
                                 .span()
                                 .unwrap()
-                                .error(format!(
-                                    "{} cannot take self by reference with a custom lifetime, use \
-                                     &self or self: Vc<Self> instead",
-                                    definition_context.function_type(),
-                                ))
+                                .error("Unexpected `self` type, use `&self` or `self: Vc<Self>")
                                 .emit();
                             return None;
                         }
-                        _ => {}
                     }
+                    // We don't validate that the user provided a valid
+                    // `turbo_tasks::Vc<Self>` here.
+                    // We'll rely on the compiler to emit an error
+                    // if the user provided an invalid receiver type
```

Comment:
For future reference, if you're moving comments like this around, IMO it's nice to re-wrap them. Until about a year ago (https://github.com/vercel/next.js/pull/68541), we were wrapping all of our comments to 80 columns due to a bug in the default configuration of a nightly rustfmt feature... It's not easy to fix everything with a codemod (the automatic line wrapping algorithm isn't 100% safe).

If you just unwrap comments and save, your editor should re-apply wrapping when it runs rustfmt.